tsdb/importer/importer.go

```diff
@@ -0,0 +1,534 @@
+				continue
+			}
+
+			_, ctime, cvalue := parser.Series()
+			if ctime == nil {
+				return 0, nil, NoTimestampError
+			} else {
+				// OpenMetrics parser multiples times by 1000 - undoing that.
+				ctimeCorrected := *ctime / 1000
+
+				var clabels labels.Labels
+				_ = parser.Metric(&clabels)
+
+				sample := tsdb.MetricSample{
+					Timestamp: ctimeCorrected,
+					Value:     cvalue,
+					Labels:    labels.FromMap(clabels.Map()),
+				}
+
+				encBuf.Reset()
+				enc := gob.NewEncoder(encBuf)
+				err = enc.Encode(sample)
+				if err != nil {
+					return 0, nil, err
+				}
+
+				advance += lineLength
+				return advance, encBuf.Bytes(), nil
+			}
+		}
+	}
+
+	// Use a streaming approach to avoid loading too much data at once.
+	scanner := bufio.NewScanner(f)
+	scanner.Split(streamSamples)
+
+	for scanner.Scan() {
+		if currentPassCount == 0 {
+			blocks = getEmptyBlocks(len(blockMetas))
+		}
+
+		encSample := scanner.Bytes()
+		decBuf := bytes.NewBuffer(encSample)
+		sample := tsdb.MetricSample{}
+		if err := gob.NewDecoder(decBuf).Decode(&sample); err != nil {
+			level.Error(logger).Log("msg", "failed to decode current entry returned by file scanner", "err", err)
+			return err
+		}
+
+		// Find what block does this sample belong to.
+		blockIndex := -1
+		if len(blockMetas) == 1 || sample.Timestamp < dbMint {
+			blockIndex = 0
+		} else if sample.Timestamp >= dbMaxt {
+			blockIndex = len(blockMetas) - 1
+		} else {
+			for blockIdx, meta := range blockMetas {
+				if meta.mint <= sample.Timestamp && sample.Timestamp < meta.maxt {
+					blockIndex = blockIdx
+					break
+				}
+			}
+		}
+
+		if blockIndex < 0 || blockIndex >= len(blockMetas) {
+			return NoBlockMetaFoundError
+		}
+		meta := blockMetas[blockIndex]
+		meta.mint = value.MinInt64(meta.mint, sample.Timestamp)
+		meta.maxt = value.MaxInt64(meta.maxt, sample.Timestamp)
+		meta.count += 1
+
+		blocks[blockIndex] = append(blocks[blockIndex], &sample)
+
+		currentPassCount += 1
+		sampleCount += 1
+		// Have enough samples to write to disk.
+		if currentPassCount == maxSamplesInMemory {
+			if err := flushBlocks(dbDir, blocks, blockMetas, logger); err != nil {
+				return err
+			}
+			// Reset current pass count.
+			currentPassCount = 0
+		}
+	}
+
+	// Flush any remaining samples.
+	if err := flushBlocks(dbDir, blocks, blockMetas, logger); err != nil {
+		return err
+	}
+
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// flushBlocks writes the given blocks of samples to disk, and updates block metadata information.
+func flushBlocks(dbDir string, blocksToFlush [][]*tsdb.MetricSample, blockMetas []*newBlockMeta, logger log.Logger) error {
+	for blockIdx, block := range blocksToFlush {
+		// If current block is empty, nothing to write.
+		if len(block) == 0 {
+			continue
+		}
+		meta := blockMetas[blockIdx]
+		// Put each sample into the appropriate block.
+		bins, binTimes := binSamples(block, meta, tsdb.DefaultBlockDuration)
```

Comment:
Yeah, that is required to actually split the samples into blocks of 2h each. In the first step, we only drop each sample into either before `dbMint`, after `dbMaxt`, or aligned with existing block. In this step, we actually divide the before `dbMint`, and after `dbMaxt` blocks into proper 2h blocks.

Reply:
why do you need to split the blocks into 2h? 

The idea is to align to the existing blocks.

so for example if you have blocks with ranges 
2000-5000,
5000-10000,
10000-20000

and if you have samples with times
2010 , 2011 .... 2100 - these produce a block with a range 2010-2100 or 2000-5000 (whichever makes the implementation easyer)

5001 , 5002 .... 5100 - these produce a block with a range 5001-5100 or 5000-10000

and so on.

if it is a new db without existing blocks can use the `tsdb.DefaultOptions.BlockRanges`

Reply:
Yes, what you're saying makes sense, and that is what happens when I have existing blocks to align with. It only splits the samples that are outside DB time limits into blocks of 2h.

Reply:
as I mentioned when you don't have existing blocks just use `tsdb.DefaultOptions.BlockRanges`

Reply:
Okay, I'll take a look at that. Should have something for you by the weekend!

Reply:
Okay, so I took a look at the constant you are referring to, and I'm already using it as `tsdb.DefaultBlockDuration`, to divide the blocks into 2h. I'm not sure if there's a place where I can use `tsdb.DefaultOptions.BlockRanges`.

Reply:
ignore the 2h split all together, just use the ranges of the blocks for an existing DB for examples `ranges int[]{2000,5000,10000}` if there is no existing DB construct these ranges from `tsdb.DefaultOptions.BlockRanges`

It might be easyer to ping me on IRC to continue the discussion.